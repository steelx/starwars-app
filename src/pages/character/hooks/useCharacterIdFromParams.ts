import {useParams} from "react-router-dom";

type CharacterIdFromParams = () => string

export const useCharacterIdFromParams: CharacterIdFromParams = () => {
    const params = useParams<{id: string}>();

    if (typeof params?.id === "undefined") {
        throw new Error('Missing character Id in params!');
    }

    return params.id;
}
