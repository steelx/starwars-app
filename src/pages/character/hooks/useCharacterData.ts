import {useCharacterIdFromParams} from "@pages/character/hooks/useCharacterIdFromParams";
import {useGetCharacterQuery} from "@store/star-wars-api/starWarsApi";
import {ICharacter} from "@store/star-wars-api/types";

interface UseCharacterDataReturn {
    data: ICharacter | undefined
    isLoading: boolean
}

export const useCharacterData = (): UseCharacterDataReturn => {
    const id = useCharacterIdFromParams()
    const { data, isLoading } = useGetCharacterQuery({id})

    return {
        data, isLoading
    }
}
