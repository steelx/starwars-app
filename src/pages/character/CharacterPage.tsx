import React from "react";
import {NavLink} from "react-router-dom";
import CharacterDetails from "./components/CharacterDetails";
import {Container} from "@shared/styles";
import {useCharacterData} from "@pages/character/hooks/useCharacterData";

const CharacterPage: React.FC = () => {
    const { data, isLoading } = useCharacterData()

    if (isLoading || !data) return null

    return <Container>
        <header><NavLink to="/">Back</NavLink></header>
        <CharacterDetails character={data} />
    </Container>
}

export default CharacterPage;
