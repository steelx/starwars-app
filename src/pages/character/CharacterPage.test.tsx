import {render} from "@testing-library/react";
import {MockedFunction} from "vitest";
import {MemoryRouter} from 'react-router-dom'
import {mockCharacter} from "@test-helpers/character-data";
import CharacterPage from "@pages/character/CharacterPage";
import {useCharacterData} from "@pages/character/hooks/useCharacterData";

vi.mock("@pages/character/hooks/useCharacterData")
vi.mock("@store/star-wars-api/starWarsApi", (module) => ({
    ...module,
    useGetCharacterFilmsQuery: vi.fn().mockImplementation(() => ({
        isLoading: true,
        refetch: vi.fn()
    }))
}))

describe("CharacterPage", () => {
    const mockUseCharacterData = useCharacterData as MockedFunction<typeof useCharacterData>;

    test("renders correctly", () => {
        mockUseCharacterData.mockImplementation(() => ({
            data: mockCharacter,
            isLoading: false
        }))

        const {asFragment} = render(<MemoryRouter><CharacterPage /></MemoryRouter>)
        expect(asFragment()).toMatchSnapshot()
    })
});
