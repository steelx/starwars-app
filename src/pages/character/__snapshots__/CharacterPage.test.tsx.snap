// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterPage > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-hLQSwg jLdJAy"
  >
    <header>
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        Back
      </a>
    </header>
    <div
      class="sc-beySPh lmGOwP"
    >
      <div
        class="sc-guDLey eYNAQj"
      >
        <h2>
          C-3PO
        </h2>
        <ul>
          <li>
            Hair Color: n/a
          </li>
          <li>
            Eye Color: yellow
          </li>
          <li>
            Height: 167
          </li>
          <li>
            Gender: n/a
          </li>
          <li>
            Homeworld: https://swapi.dev/api/planets/1/
          </li>
        </ul>
      </div>
      <div
        class="sc-dmyCSP jneFik"
      >
        <h3>
          Films Appeared In:
        </h3>
        <ul>
          <li>
            Fetching film titles...
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
